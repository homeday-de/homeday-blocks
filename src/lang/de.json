{
  "GENERAL": {
    "EDIT": "Bearbeiten",
    "SAVE": "Speichern",
    "CANCEL": "Abbrechen",
    "SAVING": "Speichern"
  },
  "FORM": {
    "EMAIL": {
      "PLACEHOLDER": "z.b. michael.schmidt@gmail.com",
      "LABEL": "Ihre E-Mail-Adresse"
    },
    "PASSWORD": {
      "LABEL": "Ihr Passwort",
      "LABEL_CONFIRM": "Bestätigen Sie Ihr Passwort"
    },
    "VALIDATION": {
      "REQUIRED": "Dies ist ein Pflichtfeld.",
      "INVALID_EMAIL": "Bitte geben Sie eine korrekte E-Mail-Adresse ein.",
      "INVALID_DATE": "Bitte geben Sie ein Datum in diesem Format ein: JJJJ-MM-TT",
      "PASSWORD_MISMATCH": "Dies stimmt nicht mit dem obigen Passwort überein",
      "PASSWORD_SHORT": "Dieses Passwort ist zu kurz",
      "PASSWORD_STRENGTH": {
        "TITLE": "Stärke",
        "LEVELS": [
          "Schwach",
          "Mittlere",
          "Stark",
          "Sehr stark"
        ]
      }
    }
  },
  "EXPAND_TEXT": {
    "TOGGLE": {
      "SHOW_MORE": "weiterlesen",
      "SHOW_LESS": "zuklappen"
    }
  },
  "GALLERY_TILES": {
    "MORE_PHOTOS": "Weitere Fotos"
  }
}
