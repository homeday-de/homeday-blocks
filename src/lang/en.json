{
  "GENERAL": {
    "EDIT": "Edit",
    "SAVE": "Save",
    "CANCEL": "Cancel",
    "SAVING": "Saving"
  },
  "FORM": {
    "EMAIL": {
      "PLACEHOLDER": "e.g. will.smith@gmail.com",
      "LABEL": "Email"
    },
    "PASSWORD": {
      "LABEL": "Password",
      "LABEL_CONFIRM": "Confirm password"
    },
    "VALIDATION": {
      "REQUIRED": "This field is required",
      "INVALID_EMAIL": "Please enter a valid email address",
      "INVALID_DATE": "Please enter a date in this format: YYYY-MM-DD",
      "PASSWORD_MISMATCH": "The entered passwords are not matching",
      "PASSWORD_SHORT": "This password is too short",
      "PASSWORD_STRENGTH": {
        "TITLE": "Strength Level",
        "LEVELS": [
          "Weak",
          "Average",
          "Strong",
          "Very strong"
        ]
      }
    }
  },
  "EXPAND_TEXT": {
    "TOGGLE": {
      "SHOW_MORE": "show more",
      "SHOW_LESS": "show less"
    }
  },
  "GALLERY_TILES": {
    "MORE_PHOTOS": "More photos"
  }
}
