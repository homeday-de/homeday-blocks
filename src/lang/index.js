// @ts-check
import de from './de.json';
import en from './en.json';

/**
 * @typedef {{ [key in keyof de]: de[key] }} Messages
 */

const langs = {
  de,
  en,
};

/**
 * Get static Homeday Blocks translation files
 *
 * @param {string} lang
 * @returns {Messages}
 */
export function getMessages(lang) {
  return langs[lang] || langs.de;
}

export default {
  getMessages,
};
