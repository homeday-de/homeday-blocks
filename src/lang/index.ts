import de from './de.json';
import en from './en.json';

export type Messages = { [key in keyof typeof de]: typeof de[key] };

const langs = {
  de,
  en,
};

/**
 * Get static Homeday Blocks translation files
 */
export function getMessages(lang: string): Messages {
  return langs[lang] || langs.de;
}

export default {
  getMessages,
};
