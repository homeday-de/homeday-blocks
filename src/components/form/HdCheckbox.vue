<template>
  <div
    :class="{
      'checkbox': true,
      'checkbox--active': isActive,
      'checkbox--disabled': disabled,
      'isChecked': isChecked,
      'hasError': !!error,
      'isUsingMouse': isUsingMouse,
    }"
    @keydown="setUsingMouse(false)"
    @mousedown="setUsingMouse(true)"
  >
    <label
      v-if="label"
      :id="`${name}--label`"
      class="checkbox__label"
      tabindex="-1"
    >{{ label }}</label>
    <input
      class="checkbox__input"
      type="checkbox"
      :name="name"
      :disabled="disabled"
      v-model="isChecked"/>
    <div
      :aria-checked="isChecked ? 'true' : 'false'"
      :aria-labelledby="label ? `${name}--label` : null"
      :tabindex="disabled ? -1 : 0"
      class="checkbox__inner"
      role="checkbox"
      @click="toggle"
      @keydown.space.enter.prevent="toggle"
      @focus="handleFocus"
      @blur="handleBlur"
    >
      <div class="checkbox__inner__box">
        <div class="checkbox__inner__box__overlay"></div>
        <div class="checkbox__inner__box__border"></div>
        <HdIcon
          :src="checkIcon"
          class="checkbox__inner__box__tick"
        />
      </div>
      <p
        v-if="innerLabel"
        v-html="innerLabel"
        class="checkbox__inner__label"
      />
    </div>
    <p
      v-if="error"
      class="field__error checkbox__error"
    >
      {{ error }}
    </p>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import { checkIcon } from 'homeday-blocks/src/assets/small-icons';

export default {
  name: 'HdCheckbox',
  components: {
    HdIcon,
  },
  props: {
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      default: '',
    },
    innerLabel: {
      type: String,
      default: '',
    },
    value: {
      type: Boolean,
      default: false,
    },
    required: {
      type: Boolean,
      default: false,
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      error: null,
      isActive: false,
      isUsingMouse: false,
      checkIcon,
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    isChecked() {
      return this.value;
    },
  },
  methods: {
    toggle() {
      this.$emit('input', !this.isChecked);
      this.$nextTick(this.validate);
    },
    handleFocus() {
      this.isActive = true;
    },
    handleBlur() {
      this.isActive = false;
      this.validate();
    },
    validate() {
      if (this.required && !this.isChecked) {
        this.error = this.t.FORM.VALIDATION.REQUIRED;
      } else {
        this.error = null;
      }
      return !this.error;
    },
    setUsingMouse(usingMouse) {
      this.isUsingMouse = usingMouse;
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'homeday-blocks/src/styles/mixins.scss';

.checkbox {
  $c: &;
  position: relative;
  margin-bottom: 26px;
  cursor: default;

  &--disabled {
    pointer-events: none;
  }

  &__input {
    display: none;
  }
  &__inner {
    display: flex;
    align-items: center;
    transition: outline 0.1s ease-in-out;
    #{$c}.isUsingMouse & {
      outline: 0;
    }
    &__box {
      position: relative;
      flex: 0 0 20px;
      width: 20px;
      height: 20px;
      cursor: pointer;
      border-radius: 3px;
      overflow: hidden;
      outline-width: 0;
      &__overlay {
        display: block;
        position: absolute;
        top:0; left:0;
        width: 100%;
        height: 100%;
        background-color: getShade($secondary-color, 110);
        border-radius: 50%;
        transform: scale(0);
        opacity: 0;
        transition: transform .3s, opacity .2s;
        .isChecked & {
          transform: scale(1.3);
          opacity: 1;
        }
        #{$c}--disabled & {
          background-color: getShade($quaternary-color, 70);
        }
      }
      &__border {
        display: block;
        position: absolute;
        top:0; left:0;
        width: 100%;
        height: 100%;
        border-radius: 3px;
        border: 2px solid $quaternary-color;
        transition: border-color .3s;
        #{$c}:hover & {
          border-color: getShade($quaternary-color, 80);
        }
        #{$c}.isChecked & {
          border-color: transparent;
        }
        #{$c}:not(#{$c}--active).hasError & {
          border-color: $error-color;
        }
        #{$c}--active & {
          border-color: getShade($secondary-color, 110);
        }
        #{$c}--disabled & {
          border-color: getShade($quaternary-color, 70);
        }
      }
      &__tick {
        position: absolute;
        top:0; left:0;
        width: 100%;
        height: 100%;
        opacity: 0;
        transition: opacity .2s;
        #{$c}.isChecked & {
          opacity: 1;
        }

        &::v-deep path {
          fill: $white;
        }
      }
    }
    &__label {
      flex: 1;
      margin: 0 0 0 $inline-s;
      text-align: left;
      @include font('text-small');

      #{$c}--disabled & {
        color: getShade($quaternary-color, 70);
      }
    }
  }
  &__error {
    #{$c}.hasError & {
      display: block;
    }

    #{$c}--active & {
      color: getShade($quaternary-color, 80);
    }
  }
  &__label {
    display: block;
    margin-bottom: 3px;
    font-size: 14px;
    line-height: 18px;
    color: $quaternary-color;
    transition: color 0.3s cubic-bezier(0.250, 0.460, 0.450, 0.940);
    pointer-events: none;

    #{$c}.hasError & {
      color: $error-color;
    }

    #{$c}:hover &,
    #{$c}--active &,
    #{$c}--active.hasError & {
      color: getShade($secondary-color, 110);
    }
  }
}
</style>
