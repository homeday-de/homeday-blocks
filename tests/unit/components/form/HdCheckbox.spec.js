import { mount } from '@vue/test-utils';
import HdCheckbox from '@/components/form/HdCheckbox.vue';
import { getMessages } from 'hd-blocks/lang';
import Vue from 'vue';

describe('HdCheckbox', () => {
  let wrapper;
  let checkboxInner;
  const activeClass = 'checkbox--active';
  const usingMouseClass = 'isUsingMouse';
  const errorClass = 'hasError';

  beforeEach(() => {
    wrapper = mount(HdCheckbox, {
      propsData: {
        name: 'Test checkbox',
      },
    });

    checkboxInner = wrapper.find('.checkbox__inner');
  });

  test('renders component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('At blur, the validation method is fired and the input element is not styled as active', () => {
    const mockedValidate = jest.fn();
    wrapper.setMethods({ validate: mockedValidate });
    checkboxInner.trigger('blur');

    expect(mockedValidate).toHaveBeenCalledTimes(1);
    expect(wrapper.classes()).not.toContain(activeClass);
  });

  test('At blur, the input is styled as error if not valid, not as error if is valid. The proper error message is also displayed', () => {
    wrapper.setProps({
      required: true,
      value: false,
    });

    checkboxInner.trigger('blur');

    expect(wrapper.vm.error).toBe(getMessages('de').FORM.VALIDATION.REQUIRED);
    expect(wrapper.classes()).toContain(errorClass);

    wrapper.setProps({
      value: true,
    });
    checkboxInner.trigger('blur');

    expect(wrapper.vm.error).toBe(null);
    expect(wrapper.classes()).not.toContain(errorClass);
  });

  test('At focus, the input element is styled as active', () => {
    checkboxInner.trigger('focus');

    expect(wrapper.classes()).toContain(activeClass);
  });

  test('On user input interaction, the proper kind of input device is detected', () => {
    wrapper.trigger('mousedown');
    expect(wrapper.classes()).toContain(usingMouseClass);

    wrapper.trigger('keydown');
    expect(wrapper.classes()).not.toContain(usingMouseClass);
  });

  test('A click on the checkboxes, emits the input event with the proper value and validates it right after', () => {
    const value = true;
    wrapper.setProps({ value });
    const mockedValidate = jest.fn();
    wrapper.setMethods({ validate: mockedValidate });

    checkboxInner.trigger('click');

    // that's why https://vue-test-utils.vuejs.org/guides/#what-about-nexttick
    return Vue.nextTick().then(() => {
      expect(wrapper.emitted('input')).toBeTruthy();

      const payload = wrapper.emitted('input')[0][0];
      expect(payload).toBe(!value);

      expect(mockedValidate).toHaveBeenCalledTimes(1);
    });
  });
});
