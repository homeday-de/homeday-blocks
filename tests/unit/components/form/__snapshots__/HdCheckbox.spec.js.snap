// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdCheckbox Accepts <html> code as innerLabel 1`] = `
<div class="checkbox">
  <!----> <input type="checkbox" name="Test checkbox" class="checkbox__input">
  <div aria-checked="false" tabindex="0" role="checkbox" class="checkbox__inner">
    <div class="checkbox__inner__box">
      <div class="checkbox__inner__box__overlay"></div>
      <div class="checkbox__inner__box__border"></div>
      <!---->
    </div>
    <p class="checkbox__inner__label">Hello <b>World!</b></p>
  </div>
  <!---->
</div>
`;

exports[`HdCheckbox renders component 1`] = `
<div class="checkbox">
  <!----> <input type="checkbox" name="Test checkbox" class="checkbox__input">
  <div aria-checked="false" tabindex="0" role="checkbox" class="checkbox__inner">
    <div class="checkbox__inner__box">
      <div class="checkbox__inner__box__overlay"></div>
      <div class="checkbox__inner__box__border"></div>
      <!---->
    </div>
    <!---->
  </div>
  <!---->
</div>
`;
